### DISCUSSION TITLE
tRPC v10 middleware implementation

### DISCUSSION THREAD

**Discussion by thisduck** [trpc/trpc#2221] [General]
In tRPC 9, we get the ability to apply middleware to every query/mutation in a router as a result of chaining queries/mutations. However it seems like in v10 we'd have to individually apply certain middleware to each query/mutation individually.

Let's say I wanted to add a logger to all the queries/mutations, it would be nice to be able to apply the middleware to the main router instead of having to import it into all the individual routers.

So in the example below (modified from the alpha middleware example), it would be nice to place middleware on to the main router so that it applies to any routes and merged routes, etc.

```typescript
import { initTRPC } from '@trpc/server';

export const t = initTRPC<{ ctx: Context }>()();

const logger = t.middleware(async ({ path, type, next }) => {
  const start = Date.now();
  const result = await next();
  const durationMs = Date.now() - start;
  result.ok
    ? logMock('OK request timing:', { path, type, durationMs })
    : logMock('Non-OK request timing', { path, type, durationMs });

  return result;
});



export const appRouter = t.router({
  foo: t.procedure.query(() => 'bar'),
  abc: t.procedure.query(() => 'def'),
  user: userRouter,
  post: postRouter, 
}).use(logger);
```

**Comment by thisduck** [comment]
Hmmm, I see. So the idea would be to define some kind of `baseProcedure` that would have the logger middleware applied. And then import baseProcedure everywhere.

I think that would suffice, please ignore. =)

**Reply by KATT** [reply]
Glad you figured it out :)

**Comment by huzaifahj** [comment]
So if we have 100 procedures split across different files and merged onto the main router, we should import and use the middleware'd procedure and use it in all those instances?

I can't see how this is sustainable.

Sure, it's nice to have granular procedure-level middleware. Again, that's not the general point of middleware and having it on a router-level is essential for any Node server framework.

**Reply by KATT** [reply]
Are the docs not clear? It's easy to create a reusable protected procedure. With this architecture, you're just not _forced_ to have all procedures in all routers in the same way.

See here for an example: https://trpc.io/docs/v10/middlewares

> ```ts
> const isAdmin = t.middleware(async ({ ctx, next }) => {
>   if (!ctx.user?.isAdmin) {
>     throw new TRPCError({ code: 'UNAUTHORIZED' });
>   }
>   return next({
>     ctx: {
>       user: ctx.user,
>     }
>   });
> });
> const adminProcedure = t.procedure.use(isAdmin);
> const adminRouter = t.router({
>   secretPlace: adminProcedure.query(() => 'a key'),
> });
> ```


Do you still hate it? I think it's a much more flexible architecture that doesn't force all endpoints in deeply nested routers to be homogenous.

If it's just that the docs are confusing to you, please edit that page and make a PR in a way that you would've understood it. 🙏 😸

**Reply by huzaifahj** [reply]
Thank you, I really appreciate the flexibility of this architecture and reusable procedures are cool but:

- `t.procedure.use(isAdmin)`: great! It's useful to be able to do this but...
- `t.router({}).use(isAdmin)` is needed too

Hopefully you are getting my point? This discussion was started with a suggestion (the same as mine) but was ended with a workaround. I would like to continue discussing the suggestion.

**Reply by KATT** [reply]
You can define an `adminProcedure` and just use that locally in the admin router. Call it something short and it's the same thing as using `t.procedure`

**Reply by KATT** [reply]
This is how your admin router would likely look like:

```ts
import { adminProcedure, t } from '~/server/trpc';

export const adminRouter = t.router({
  secret: adminProcedure.query(() => 'a key'),
  secret2: adminProcedure.query(() => 'a key'),
});
```

It wouldn't be too much different to an API that you want which would be something like:

```ts
import { isAdmin, t } from '~/server/trpc';

export const adminRouter = t.router
  .use(isAdmin)
  .procedures(procedure => ({
    secret: procedure.query(() => 'a key'),
    secret2: procedure.query(() => 'a key'),
  });
```


...  it's way more modular to have the authz checks etc straight on the procedure rather than on the router as you can have exceptions & it generally plays nicer together with adding sub-routers etc

**Reply by huzaifahj** [reply]
Thank you, understood

**Reply by rhyek** [reply]
@KATT completely missed the point here. if you have an admin namespaced subrouter, it makes complete sense to have the ability to apply some security middleware to all procedures within it to avoid the margin of error where a developer could forget to use the correct procedure type. think 10s or 100s of procedures within a protected router with common security checks.

**Reply by huzaifahj** [reply]
@rhyek I agree with you. Many users of tRPC within this discussion have raised ergonomics and security concerns which are being ignored by a framework that prides itself on at least one of those.

**Comment by nosuke23** [comment]
In the case of chaining middleware,  we want to define each middleware separately without `.use()` like `.middleware()`. I think procedures should have `middleware()` function.

```ts
const isAuthed = trpc.middleware(({ next, ctx }) => {
  ...
  return next({
    ctx: {
      user: ctx.session.user,
    },
  })
})
```

```ts
const withAuthMW = trpc.procedure.use(isAuthed).middleware(...)
```

instead of

```ts
const authedProcedure = trpc.procedure.use(isAuthed).use
type WithAuth = Parameters<typeof authedProcedure>[0]

const withAuthMW: WithAuth = ...
```

It's better if middleware can update other props.

```ts
const isAnyWithInputValidation = trpc.procedure.input(z.object({foo: z.string()})).middleware(({input, next}) => {
  const { foo } = input
  ...
  return next({...})
})

const anyProcedure = trpc.procedure.use(isAnyWithInputValidation).input(
  z.object({
    foo: z.string(),
    bar: z.number()
  })
).query(({input}) => {
  const { foo, bar } = input
  ...
})
```

**Comment by karmux** [comment]
I totally agree with @huzaifahj that
```
t.router({}).use(isAdmin)
```
is very much needed and I think it's a huge regression.

tRPC examples are very primitive and there it's not a big problem that can only add middleware per procedure. You have router and procedure visually in the same place.

In real life bigger projects you would put queries and mutations to separate files and then you would have to make sure that you use correct procedure in these files since it's not clear to which router they belong to. It's easy to make a mistake of putting a public procedure under admin router.

**Reply by lucashfreitas** [reply]
@karmux  exactly, I am facing this problem in almost every TRPC application I develop. Usually I have 100+ routes and some of them have different middleware/procedures (admin, group level, etc). It would be much easier to define this globally at a route level than repeating that 100 times.

Easy to do a mistake and forget to apply a procedure, or apply the wrong one - that's not too bad since we have tests for that, but would be a game changer if we could apply a middleware/procedure to a group/many route items in the route level at once. 

I **LOVE** the current implementation and flexibility, but would also be nice to define it just once at the route level than having to do it many times again and again. 

This is not a new concept and some web frameworks use that strategy of defining middleware at a route level. I imagined something like this.

<img width="506" alt="image" src="https://user-images.githubusercontent.com/11416392/224916177-70b243ac-44ba-43e1-affa-3170a1343a44.png">

**Reply by KATT** [reply]
What's the problem with using [base procedures](https://trpc.io/docs/procedures#reusable-base-procedures)? They're more composable and doesn't enforce behaviour of a procedure just because a procedure is within a certain router.

If you really feel like you **need** a protected router wrapper like this, you can write one. Here's an example to get you started:

```ts

function protectedRouter<TCallback extends (proc: typeof protectedProcedure) => Record<string,AnyProcedure>>(callback: TCallback) {
  return t.router(callback(protectedProcedure));
}

const user = protectedRouter(proc => ({
  whoami: proc.query(opts => {
    console.log('your user id is', opts.ctx.session.user.id)
  }),
}))
```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAb2AO2DAKgJQAoGEA0cWeAolFNIQILICe2FAxgKYAmArlMwL5wBmFEHABEAARhQwjAPQBnZlABuC4QChVjCMlnx4AXjgo0xXADpNyGMwAeMADwJVcOPNmzgWgPwAuRE+dw7PJQvo4BAcCsvjpQKADm-s7c-sncAHwAFACU5lwAhlbZANzq6hY6cGAUVoxWrAwQLBxccAYwplWNbJzMpkHMGRkQYDCyWa1pfs7AfHAZAITDo+Yw1qau7lrjYQEwABYUAO5wyMzHJmQUUBk74ZqszL7CAKoAclTP6AASAPKYAJIALRIABE1OFuFkUupnFwYJxkHAlrJTKdbDdEnBatZQpjnBsPMhfMiVmsCVp8JjkkkoapIeo+OxkLVCZVqsxamxMBB2FYoHZ0Lg8gAbYUAIzyjAA1nAbFZkKxZHNOoxfDBaGBmBBZp0anUGk0euM9JNMBzoKw7DF4vgaPQmN0uGlMowReLJVLfIK3RLpdt-HCEXB2hReQoMq7Rb6pRldRz9Q7mswslkSskNFoKv0oK02RA9VyeXzY0wJnMdoc9hA8iBgL4VaYAI7sBS0IYjJUmqYBcoQYW9YUQOIZADktB5OezhlYhlkI8IJOx62YbkJfWCpkiUJplMhtIz2ngeTAYG5YZzbVMoeLO2zu6yQA


You can choose not to expose e.g. unprotected procedures to the rest of your app so you can only use your router-helpers.

**Reply by davidtjones02** [reply]
> What's the problem with using [base procedures](https://trpc.io/docs/procedures#reusable-base-procedures)? They're more composable and doesn't enforce behaviour of a procedure just because a procedure is within a certain router.
> 
> If you really feel like you **need** a protected router wrapper like this, you can write one. Here's an example to get you started:
> 
> ```ts
> function protectedRouter<TCallback extends (proc: typeof protectedProcedure) => Record<string,AnyProcedure>>(callback: TCallback) {
>   return t.router(callback(protectedProcedure));
> }
> 
> const user = protectedRouter(proc => ({
>   whoami: proc.query(opts => {
>     console.log('your user id is', opts.ctx.session.user.id)
>   }),
> }))
> ```

Thanks for sharing this @KATT, i'm trying to implement this in V11 but am getting this TS error

```typescript
Property 'useQuery' does not exist on type 'DecoratedQuery<{ input: any; output: any; transformer: true; errorShape: { message: string; data: { error_id: string; zod_error: typeToFlattenedError<any, string> | null; }; code: TRPC_ERROR_CODE_NUMBER; } | { ...; }; }> | DecoratedMutation<...> | { ...; }'.
  Property 'useQuery' does not exist on type 'DecoratedMutation<{ input: any; output: any; transformer: true; errorShape: { message: string; data: { error_id: string; zod_error: typeToFlattenedError<any, string> | null; }; code: TRPC_ERROR_CODE_NUMBER; } | { ...; }; }>'.ts(23
```

The endpoint does in fact work, as in I can do something like this 

```typescript
import { trpc } from "@/features/trpc";

const Test = () => {
  const {data} = trpc.admin.test.useQuery();

  console.log(data);

  return <>div</>;
};

export default Test;
```

But the typescript for it seems to have problems on the frontend. No errors appear on the backend. 

Basic backend setup

```typescript
export const authenticatedProcedure = t.procedure.use(
  t.middleware(({ ctx, next }) => {
    return next({
      ctx: {
        ...ctx,
        user_id: 1235,
      },
    });
  })
);

export function authenticatedRouter<
  TCallback extends (
    authenticated: typeof authenticatedProcedure
  ) => Record<string, AnyTRPCProcedure>,
>(callback: TCallback) {
  return t.router(callback(authenticatedProcedure));
}

const AdminRouter = authenticatedRouter((authenticated) => ({
  test: authenticated.query(({ ctx }) => {
    console.log(ctx.user_id); // this works! and returns 1235
    return { success: true };
  }),
}));

export default AdminRouter;
```

On version 10 I copy closer to what you've done above, and set it as `AnyProcedure` instead of `AnyTRPCProcedure`. But on the frontend I just get

`Property 'useQuery' does not exist on type 'never'.ts(2339)`. Again, the query works and I do not have any type errors or problems with other TRPC endpoints. I can do `trpc.admin` which is declared on the highest level router but it won't grab the individual endpoints
```typescript
export const appRouter = router({
  admin: AdminRouter,
// other routers declared here
});
```


What do I need to do differently? Thanks for the help.

**Reply by lanternlogic** [reply]
Hey @davidtjones02 did you end up figuring this out?

**Comment by AdventureBeard** [comment]
Sorry to gravedig, but just wanted to add: not being able to apply middleware to a namespace is a huge miss, IMO. This will prevent me from migrating to V10 just from the sheer amount of present and future maintenance this change creates. My project is designed around namespaces with certain rules, such that routes added to particular routers automatically inherit all those rules. Very easy to move fast and not break anything by just putting a route in the right place.

I understand that separating routing from middleware logic is probably the *by-the-books* Separation of Concerns way to do this, but imo is a big step back in actual DX. It results in a ton of duplicate code, and now every added route has a fresh chance to break auth rules by missing a middleware/pipe wrapper.

**Comment by willisplummer** [comment]
I'm just adopting tsrpc and just discovered that the ability to make an entire router password protected existed in 9 and not 10. This was pretty surprising to me and feels like a miss - I get that it is possible to use protectedProcedure when declaring each procedure but like AdventureBeard said, it feels like there's increased risk of accidentally using publicProcedure in one place and leaving an API exposed

**Comment by ahkhanjani** [comment]
Since the discussion is still open,

In my case I have a router subtree for a specific feature that now needs to check for a feature flag for the entire route. Just adding that at the router level would make it much more predictable.

It is definitely doable with a base proc as @KATT mentioned. But in my case:

1. I already have extended procs for role based routes. They are extended from a `protectedProcedure`.
2. Adding the feature flag layer would need the nodes to now point to this new proc.
3. Later when the feature flag is no longer needed, they need to point to the `protectedProcedure` like before.

It's like adding a node in the middle of a linked list and then removing it. Whereas with a router level middleware, we're just putting a wall/door in between. No need to move pieces anymore.

Depending on the size of the tree and the complexity of the procedures it can make a big difference.

**Comment by lanternlogic** [comment]
Hey! Any further movement on this? I have just started with tRPC and having subrouter-level middleware was how I envisioned setting up my router. Hearing that tRPC doesnt have this feature doesnt make sense to me, is there any workaround? Or has anyone found a way to circumvent this

**Reply by ahkhanjani** [reply]
> Hey! Any further movement on this? I have just started with tRPC and having subrouter-level middleware was how I envisioned setting up my router. Hearing that tRPC doesnt have this feature doesnt make sense to me, is there any workaround? Or has anyone found a way to circumvent this

Hi. I recommend going with tRPC 11 and [this](https://trpc.io/docs/server/middlewares#concat) thing.

**Reply by lanternlogic** [reply]
Hey, as far as I'm aware the link you shared is for extending middlewares on _procedures_ rather than on a subrouter.

I hope that it's not just my beginners understanding with tRPC but it seems this link's middleware docs still doesn't allow me to have an api where 
- http://basedomain/trpc/**subrouter**/procedure1
- http://basedomain/trpc/**subrouter**/procedure2

where **subrouter** itself has the middlewares, instead chaining each procedure1 or procedure2.

**Reply by ahkhanjani** [reply]
My apologies. I misread your comment. Yes you are right. I'm not aware of a method of achieving this still.

**Reply by KATT** [reply]
If someone sponsors trpc with $100 I'll happily spend time doing a working version of the above

**Reply by davidtjones02** [reply]
> If someone sponsors trpc with $100 I'll happily spend time doing a working version of the above 

@lanternlogic would you be willing to split half/half? i'd be happy to help fund! 

alternatively @KATT are you able to point us in some directions so we can help contribute to a PR?

**Reply by lanternlogic** [reply]
@davidtjones02 yep! I'm actually totally up for a 50/50.
Let us know how we can pay

**Reply by ahkhanjani** [reply]
I am unable to use any international payment methods due to where I live. Thank you from a community member you guys!

> Let us know how we can pay

There should be a sponsorship button on the main page of the repository.

**Reply by davidtjones02** [reply]
Thanks @ahkhanjani ! I've sent $50 @KATT @lanternlogic  🙌

**Reply by lanternlogic** [reply]
Sent ! @KATT @davidtjones02

**Reply by KATT** [reply]
Happy with something like what's outlined here? https://github.com/trpc/trpc/discussions/2221#discussioncomment-9918838

I'm on a work trip this week but will see if I find some time in the next few days. Thanks for the money 🙏

**Reply by lanternlogic** [reply]
Just had a look at that code snippet above - mentally parsing that Typescript humbles me a bit, but i think I understand the idea there. Essentially you are wrapping the protectedRouter around the procedures such that all executions of queries/mutations within it like `whoami` have already executed the logic in `protectedProcedure`?

I presume that will have full type safety that we know n love with tRPC? Looks great then.

Thank you!!!

**Reply by ahkhanjani** [reply]
> Happy with something like what's outlined here? [#2221 (reply in thread)](https://github.com/trpc/trpc/discussions/2221#discussioncomment-9918838)

I think instead of:

```ts
function protectedRouter<TCallback extends (proc: typeof protectedProcedure) => Record<string,AnyProcedure>>(callback: TCallback) {
  return t.router(callback(protectedProcedure));
}

const user = protectedRouter(proc => ({
  whoami: proc.query(opts => {
    console.log('your user id is', opts.ctx.session.user.id)
  }),
}))
```

an easier path would be to make the wrapper accept a `TRPCRouterRecord` (a regular object containing routes) and apply `unstable_concat` to each one. Like this:

```ts
const user = routerMiddleware({
  whoami: proc.query(opts => {
    console.log('your user id is', opts.ctx.session.user.id)
  }),
})
```

`whoami` endpoint here would have `routerMiddleware` which is a regular middleware automatically concatenated.

**Reply by KATT** [reply]
The second example isn't possible unfortunately, but the first one is, see https://github.com/trpc/trpc/pull/6070/files

Happy with this?

**Reply by KATT** [reply]
FWIW, I still **do not recommend doing this**; it's nicer to group your routers _per domain_ rather than based on what the caller can do.

In the same PR, I just added another reference how you can validate your router based on your conventions at runtime by leveraging `.meta()` in different builders

https://github.com/trpc/trpc/blob/997d2cf23bfd39272ed87f5a4d4583b7f61059e5/packages/tests/server/regression/issue-2221-router-factory-fn.test.ts#L183-L241

You can take this further and have e.g. `orgProcedure`s that extends `authedProcedure`s. Reference:

- https://x.com/alexdotjs/status/1839265630983213432
- https://tsplay.dev/mA6KQW

**Reply by ahkhanjani** [reply]
Thank you @KATT.

> The second example isn't possible unfortunately

I can see why now.

> In the same PR, I just added another reference how you can validate your router based on your conventions at runtime by leveraging `.meta()` in different builders

Aha! I've been staring at my screen for half an hour and my light bulb just lit up. I see what you're doing.
Honestly with the limitations of this router based middleware not cascading I'm not so sure about this either. Although I'm certain there must be some trick we can use to achieve that but I don't think it'll be worth it. A tRPC design limitation maybe?

> ```ts
> const user = protectedRouter(proc => ({
>   whoami: proc.query(opts => {
>     console.log('your user id is', opts.ctx.session.user.id)
>   }),
> }))
> ```

Doing something like this when different routes are defined separately can be a problem:

```ts
const router0 = proc0Wrapper((proc0) => ({
  router1: proc1Wrapper((proc1) => ({
    whoami: proc0.unstable_concat(proc1).query(...), // Looks just as painful as not using router based middleware.

    router2: router2,
  })
}));

const router2 = {
  ... // No access to procs 0 and 1.
};

// Solution:
const router2 = (proc) => ({
  ...
});

// Then, inside router0:
{
  router2: router2(proc0.unstable_concat(proc1)),
};
```

Have I understood the concept correctly? If yes this just looks like a bad idea to me.

**Comment by sebbean** [comment]
ty for all your hard work @KATT 

just joining the trpc gang last few months, just upgrading from 10 to 11 (ty for your work with wsClient connectionParameters verry usefule https://github.com/trpc/trpc/pull/5839)

figure this is as god as any place to ask this question (happy New Year btw!)

How would you suggest deciding between adding to ctx vs meta in the middleware?

i'm specifically referring to auth but i guess it can be expanded more generally

for instance i'm using createContext to populate things like token and user, but should i be using meta instead?

**Reply by sebbean** [reply]
Ok i think i answered my own question by looking over your previously posted links

you can chain the meta to the end of a procedure to flag it for auth logic or not when implementing

https://github.com/trpc/trpc/blob/997d2cf23bfd39272ed87f5a4d4583b7f61059e5/packages/tests/server/regression/issue-2221-router-factory-fn.test.ts#L30-L46

**Reply by KATT** [reply]
See https://github.com/trpc/trpc/discussions/5196#discussioncomment-7913438 for an advanced version of this

**Reply by KATT** [reply]
& https://x.com/alexdotjs/status/1839265630983213432

**Comment by jaime-linvix** [comment]
You know what? That's it, everyone's wrong, express, fastify, hono, koa, my mother, everyone. And tRPC is right. Oh come on, by god's sake, wouldn't it just make sense letting people add middlewares to both routers and procedures and let everyone do it their way?

**Reply by KATT** [reply]
Thanks for the passionate feedback! We certainly made some deliberate choices with tRPC's middleware structure, trying to balance flexibility with that sweet, sweet type-safety.

But hey, the beauty of open source means you're always free to use a competitor, build your own library, or contribute to tRPC with code or financially to help us do the thing you want. 

Good luck on your next adventure, we'll certainly miss your many constructive contributions dearly.

**Reply by jaime-linvix** [reply]
Thanks for your answer, mate, but you kindly complained without presenting any counter argument. Your creation is absolutely great, and that's precisely the reason I'm mad with this change, but changing into this direction makes 0 sense to me, and presumably to the majority of devs I've seen talk about this change. I'm about to switch to hono RPC to try something new. If I like it, then I'll stick to it.

**Comment by jwilm** [comment]
Hey folks,

Wanted to add some feedback here as someone who has been thrown into a rather large tRPC project (multiple hundreds of procedures).

My use case is simple. I want to add a tracing middleware to our top level router which will create a standard span with standard otel semantic attributes. I would like to write that code in one place rather than having to modify the definitions of hundreds of procedures. Further, if it ever needs to be refactored, I would hate to have to go make that change in hundreds of places.

While Jaime's feedback was certainly on the snarky end, I think their point has merit that coming from basically any other HTTP/RPC ecosystem where you are able to define centralized middlewares results in a jarring experience in tRPC.

Seems like myself and most others in this thread would benefit greatly from the ability to define router-level middlewares.

Thanks for reading (and for tRPC!)
